viewing_party/party.py

```diff
@@ -1,23 +1,175 @@
+# -----------------------------------------
 # ------------- WAVE 1 --------------------
+# -----------------------------------------
 
 def create_movie(title, genre, rating):
-    pass
+    movie = {}
+
+    if not title or not genre or not rating:
+        return None
+    else:
+        movie["title"] = title
+        movie["genre"] = genre
+        movie["rating"] = rating
+
+    return movie
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+
+    return user_data   
+
+def watch_movie(user_data, movie):
+    # If the movie is the watchlist, it's moved from the watchlist 
+    # to the watched
+    for i in range(len(user_data["watchlist"])):
+        if movie == user_data["watchlist"][i]["title"]: 
+            user_data["watched"].append(user_data["watchlist"][i])
+            del user_data["watchlist"][i]
+
+    return user_data    
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0.0
+    
+    count = 0
+    sum_ratings = 0
+
+    for movie in user_data["watched"]:
+        sum_ratings += movie["rating"]
+        count += 1
+
+    avg_rating = sum_ratings/count
+
+    return avg_rating
+
+def get_most_watched_genre(user_data):
+    genres_count = {}
+    
+    # make a dictionary where genre is key and 
+    # how many times genre watched is value
+    if not user_data["watched"]:
+        return None
+    else:
+        for movie in user_data["watched"]:
+            genres_count[movie["genre"]] = genres_count.get(movie["genre"], 0) + 1
+    
+    most_watched_genre = None
+    times = 0
+    # iterate through genres_count dict
+    # find genre with highest count
+    for genre, count in genres_count.items():
+        if count > times:
+            most_watched_genre = genre
+            times = count
+    
+    return most_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    unique_watched = []
+    set_user_mov_titles = set()
+    set_mov_friends_titles = set()
+
+    # create set of friends movies titles
+    for fr_watched in user_data["friends"]:
+        for mov in fr_watched["watched"]:
+            set_mov_friends_titles.add(mov["title"])
+        
+    # create set of user movies titles
+    for movie in user_data["watched"]:
+        set_user_mov_titles.add(movie["title"])

```

Comment:
I love the move here to create sets of movies that hold the user's movies and the friends' movies. This makes it so much easier to find the unique movies as opposed to trying to do it all in place. I also love the move to go ahead and only make sets of titles, since we know we can easily get rid of any duplicates that way and sets don't work with full dictionaries!